### PR TITLE
chore(tests): Basic flow decider tests [NET-577]

### PR DESCRIPTION
**Pull request by kmd-fl**
~~Note that atm the code doesn't look nice and isn't structured well since it's wip.~~

### CODE REVIEW

File: src/tests/decider-distro-tests-rs/Cargo.toml [hunk 1]
```diff
@@ -7,14 +7,25 @@ edition = "2021"
...
-system-services = { git = "https://github.com/fluencelabs/nox.git", branch = "plugable-system-services" }
+created-swarm = { git = "https://github.com/fluencelabs/nox.git", branch = "fix-rescheduling-spells" }
+connected-client = { git = "https://github.com/fluencelabs/nox.git", branch = "fix-rescheduling-spells" }
+system-services = { git = "https://github.com/fluencelabs/nox.git", branch = "fix-rescheduling-spells" }
+log-utils = { git = "https://github.com/fluencelabs/nox.git", branch = "fix-rescheduling-spells" }
 fluence-app-service = { version = "=0.26.4-feat-wasm-backend-interface-update-b3cf050-636-1.0", registry = "fluence" }
 
+fluence-keypair = "0.10.3"
+
+
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/Cargo.toml:18]:
```suggestion
```

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 2]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+        chain_matcher_addr: "0x0".to_string(),
+        chain_workers_gas: 210_00,
+        chain_wallet_key: "0xfdc4ba94809c7930fe4676b7d845cbf8fa5c1beae8744d959530e5073004cf3f"
+            .to_string(),
+    };
+    // decider will run once
+    make_distro(config, decider_settings)
+}
+
+async fn execute(
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:130]: WDYT, should that be here or in nox's test utils?

I'm thinking about whether this and some other functions from here would be useful in tests in other spells, eg in fluencelabs/spell.

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 3]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+            "config" => json!(trigger_config),
+        },
+    )
+    .await
+}
+
+fn enable_decider_logs() {
+    let namespaces = vec![
+        "run-console=debug",
+        // "spell_event_bus=trace",
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:165]: let's clean up that a little bit. I think it's OK to have commented lines, but I'd wish them more consistently formatted

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 4]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+            )
+        "#,
+            r#"status"#,
+            hashmap! {
+                "worker" => json!(worker_id),
+            },
+        )
+        .await
+        .unwrap();
+        assert!(!result.is_empty());
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:269]: add an error message pls

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 5]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+        let strings = serde_json::from_value::<StringListValue>(result[0].clone()).unwrap();
+        assert!(strings.success);
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:271]: ditto

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 6]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+        if !strings.strings.is_empty() {
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:272]:
```suggestion

        assert!(!result.is_empty());
        let strings = serde_json::from_value::<StringListValue>(result[0].clone()).unwrap();
        assert!(strings.success);

        if !strings.strings.is_empty() {
```

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 7]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+            #[derive(Deserialize, Debug)]
+            struct State {
+                state: String,
+            }
+            let last_status = strings.strings.last().unwrap();
+            let state = serde_json::from_str::<State>(last_status).unwrap();
+            if state.state != "INSTALLATION_IN_PROGRESS" {
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:279]: Wouldn't that fail if called too early, when status is not yet `INSTALLATION_IN_PROGRESS`?

Maybe have a match over `state.state` and accept all non-error statuses?

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 8]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+    }
+}
+
+async fn wait_decider(client: &mut ConnectedClient) {
+    let decider_stopped = client
+        .receive_args()
+        .await
+        .wrap_err("wait decider")
+        .unwrap();
+    assert_eq!(decider_stopped[0].as_str().unwrap(), "done");
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:294]: message could be a bit less terse, I think :)

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 9]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+        "#,
+        "cid",
+        hashmap! {
+            "worker_id" => json!(worker_id),
+        },
+    )
+    .await
+    .unwrap();
+    let result = serde_json::from_value::<StringValue>(result.remove(0)).unwrap();
+    assert!(!result.absent, "worker-spell doesn't have worker_def_cid");
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:314]: now that's a great message!

File: src/tests/decider-distro-tests-rs/tests/test.rs
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:1]: I think it might make sense to separate that file into several.

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 10]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
         assert!(names.contains(&alias2), "{alias2} service is not installed");
     }
 }
+/// Test the basic flow
+///
+/// 1. *Decider* asks the last block of the chain from which to start polling
+///    The block number is `0x10`
+///
+/// 2. *Decider* asks for the logs from the block `0x10` to the `0x10 + 500` blocks (range configured in the connector)
+///    We return the logs with pre-defined CID of the url-downloader app. This step required IPFS node with the app.
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:542]: What does "requires IPFS node with the app" mean?

I'd read it as `"Downloading this CID requires an available IPFS instance"`. Is that correct?

```suggestion
///    We return the logs with pre-defined CID of the url-downloader app. This step requires IPFS node with the app.
```

Comment (folex) [reply src/tests/decider-distro-tests-rs/tests/test.rs:542]: or does "with the app" mean `"an IPFS instance with url-downloader app package pre-uploaded to it"`?

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 11]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+///    Here, as a part of testing the basic flow, we check that the app was deployed correctly:
+///    - `srv.list` on the worker returns the `worker-spell` spell and the `url-downloader` service
+///
+/// 8. *Decider* polls updates for already existing deals.
+///    Check that *Decider* doesn't try to find updates for the new deal.
+///
+/// NOTE: This test REQUIRES an IPFS node to be up and have the url-downloader app uploaded.
+/// TODO: provide the app in the tests resources
+/// TODO: checks that `errors` are empty
+///
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:572]: 🔥

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 12]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+                    log.topics[1].as_str(),
+                )])
+            }
+            "eth_sendRawTransaction" => {
+                // how to check correctness of the subnet registration?
+                json!("0x55bfec4a4400ca0b09e075e2b517041cd78b10021c51726cb73bcba52213fa05")
+            }
+            "eth_getTransactionCount" => json!("0x1"),
+            "eth_gasPrice" => json!("0x3b9aca07"),
+            _ => panic!("unexpected method: {}", method),
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:597]:
```suggestion
            _ => panic!("mock http got an unexpected rpc method: {}", method),
```

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 13]
```diff
@@ -48,15 +75,402 @@ pub fn make_distro() -> PackageDistro {
...
+            if state.state != "INSTALLATION_IN_PROGRESS" {
+                assert_eq!(state.state, "INSTALLATION_SUCCESSFUL");
+                break;
+            }
+        }
+        tokio::time::sleep(timeout_per_try).await;
+    }
+}
+
+async fn wait_decider(client: &mut ConnectedClient) {
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:288]:
```suggestion
async fn wait_decider_stopped(client: &mut ConnectedClient) {
```

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 14]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+            )
+        "#,
+        "counter deals",
+        hashmap! {},
+    )
+    .await
+    .unwrap();
+
+    // Check that decider was really run
+    let counter = serde_json::from_value::<U32Value>(result.remove(0)).unwrap();
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:638]: i'd move `result.remove(0)` to a separate variable with a telling name. `result[0]` is hard to follow

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 15]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+        )
+        .await
+        .unwrap();
+        serde_json::from_value::<Vec<ServiceInfo>>(result[0].clone()).unwrap()
+    };
+
+    let test_app_1 = TestApp::test_app1();
+    let worker_spell = worker_service_list
+        .iter()
+        .find(|info| info.aliases.contains(&"worker-spell".to_string()));
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:719]: `& .to_string()` looks strange, sure it won't take `&'static str`?

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 16]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+    assert!(worker_spell.is_some(), "no worker-spell on the worker");
+    let worker_spell = worker_spell.unwrap();
+    assert_eq!(
+        worker_spell.service_type, "spell",
+        "worker-spell is not a spell"
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:724]:
```suggestion
        "worker-spell is not a spell, but {}", worker_spell.service_type
```

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 17]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+        .iter()
+        .find(|info| info.aliases.contains(&test_app_1.services_names[0]));
+    assert!(
+        test_service.is_some(),
+        "no test service on the worker from a deal"
+    );
+    let test_service = test_service.unwrap();
+    assert_eq!(
+        test_service.service_type, "service",
+        "test service is not a service"
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:747]:
```suggestion
        "test service is not a service, it is {}", test_service.service_type
```

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 18]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+        );
+        let (app, block) = result.unwrap();
+
+        let cid = get_worker_app_cid(&mut client, &deal.worker_id).await;
+        assert_eq!(cid, app.cid, "wrong cid");
+        let deal_state = get_deal_state(&mut client, &deal.deal_id).await;
+        assert_eq!(deal_state.left_boundary, to_hex(block), "wrong saved state");
+    }
+
+    server.shutdown().await;
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:898]: is it possible to do in server's destructor? So we never forget to do that.

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 19]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+/// Test deal deployment in different *Decider* runs
+///
+/// Plan
+/// 1. *Decider* installs a deal.
+///     This process already was checked.
+/// 2. *Decider* installs another deal.
+///    Check exactly what can be changed from the installation of another deal
+///    a. `joined_deals` list contains both deals (check that the list is not overwritten)
+///    b. state of the deal (stored by `deal_id`)
+/// 3. Install a
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:910]: seems like part of the comment was deleted :(

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 20]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+        swarms[0].multiaddr.clone(),
+        Some(swarms[0].management_keypair.clone()),
+    )
+    .await
+    .unwrap();
+
+    update_decider_script_for_tests(&mut client, swarms[0].tmp_dir.clone()).await;
+    // Initial run for installing the first deal
+    update_config(&mut client, &oneshot_config()).await.unwrap();
+    {
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:944]: Why a separate scope? Seems like `for { }` would be enough

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 21]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+            &mut client,
+            r#"
+            (seq
+                (call relay ("decider" "get_string") ["last_seen_block"] last_seen)
+                (seq
+                    (call relay ("decider" "list_get_strings") ["joined_deals"] deals)
+                    (call relay ("worker" "list") [] workers)
+                )
+            )
+        "#,
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:1024]: indentation

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 22]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+                "eth_getTransactionCount" => json!("0x1"),
+                "eth_gasPrice" => json!("0x3b9aca07"),
+                _ => panic!("unexpected method: {}", method),
+            };
+            send_response.send(Ok(response)).unwrap();
+        }
+    }
+    wait_decider(&mut client).await;
+
+    let (last_seen, deals, mut workers) = {
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:1013]: Maybe DRY this block?

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 23]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+    assert!(
+        deals.is_empty(),
+        "since the registration failed, the deal must not be joined"
+    );
+
+    update_config(&mut client, &oneshot_config()).await.unwrap();
+    {
+        // Reqs: blockNumber, getLogs, gasPrice, getTransactionCount and sendRawTransaction
+        for _ in 0..5 {
+            let (method, params) = recv_request.recv().await.unwrap();
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:1307]: What if decider breaks, will it hang forever?

Does decider ever breaks in these tests, ever?

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 24]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+                    assert!(log.from_block <= BLOCK_NUMBER);
+                    assert!(log.to_block <= BLOCK_NUMBER);
+                    json!([TestApp::log_test_app1(
+                        DEAL_ID,
+                        BLOCK_NUMBER,
+                        log.topics[1].as_str()
+                    ),])
+                }
+                "eth_sendRawTransaction" => {
+                    // TODO: how not to wait for the registration if Decider failed?
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:1321]: That's a very interesting question, indeed. I don't know how.

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 25]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+        }
+    }
+    wait_decider(&mut client).await;
+    let deals = get_joined_deals(&mut client).await;
+    assert!(!deals.is_empty(), "the deal must be joined after fail")
+}
+
+/*
+// Update doesn't work and we don't know how it will work in future
+// Maybe, make this test ALWAYS fail to remind that this doesn't work?
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:1338]: I think we can implement this test. Return a `DealChanged` log from RPC mock, and check that all the logic works.

Having this feature working _almost_ correctly is enough. Fundamental things won't change, and for the details we can tweak as they appear.

Same holds for Deal removal.

File: src/tests/decider-distro-tests-rs/tests/test.rs [hunk 26]
```diff
@@ -119,3 +533,1006 @@ async fn test_decider_installed() {
...
+/// 3. `0x10` -- again the same number in not very realistic case when Decider is too fast and the chain is too slow
+///            Decider shouldn't move it's left boundary anywhere, so
+///            saved last_seen_block should be `0x10`
+///            In the eth_getLogs request: we can check that fromBlock is `0x11`
+///
+/// 4. `0xffffff` -- big number which Decider shouldn't be able to process during one iteration
+///            Decider should move to some boundary configured boundary, which is 500 blocks, so
+///            saved last_seen_block should be `0x205`
+///            In the eth_getLogs request: we can check that fromBlock is `0x11`
+///            Note: the test depend on the knowledge that the range is 500 blocks,
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test.rs:1470]: Sounds more and more that this number should be part of `init data`. WDYT?

File: src/tests/decider-distro-tests-rs/tests/test_rpc_server.rs [hunk 27]
```diff
@@ -0,0 +1,151 @@
...
+    let recv_resp = Arc::new(Mutex::new(recv_resp));
+
+    let process_http_request =
+        async move |mut req: Request<Body>| -> Result<Response<Body>, Infallible> {
+            let raw_body = req.body_mut();
+            let mut buf = body::aggregate(raw_body).await.unwrap();
+            let body = buf.copy_to_bytes(buf.remaining()).to_vec();
+            let raw_request = serde_json::from_slice::<Value>(&body).unwrap();
+
+            let response = if raw_request.is_array() {
```
Comment (folex) [inline src/tests/decider-distro-tests-rs/tests/test_rpc_server.rs:55]: maybe separate these two parts of the `if` in two functions?

**Review by folex — APPROVED** [review]

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 26 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs:1197]: probably, but it feels like the idea for a separate refactoring pr

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 19 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs]: don't remember what I wanted to write here :see_no_evil: 
edited

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 18 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs:471]: I tried but it didn't work as expected

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 15 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs:290]: No, `contains` wants `&String

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 10 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs]: Yep. It means that in CI we need to have IPFS node with the specific uploaded services.

File: src/tests/decider-distro-tests-rs/tests/test.rs
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs:1]: I moved out utils stuff. Do you want more separation?

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 2 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs]: would be nice, not sure if exactly this interface is good for a lib though

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 22 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs:580]: I wonder what this block should be called. I can't see how to unite this concepts

File: src/tests/decider-distro-tests-rs/tests/test.rs
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs:1]: I also moved out deployment tests to a different file. Now the structure is the following:
```
tests
├── test_deploy.rs
├── tests_basic.rs
└── utils
    ├── mod.rs
    └── test_rpc_server.rs
```

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 26 above)
Comment (folex) [reply src/tests/decider-distro-tests-rs/tests/test.rs]: can you create a task for that plz?

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 25 above)
Comment (folex) [reply src/tests/decider-distro-tests-rs/tests/test.rs]: can you create a task for that plz?

File: src/services/chain-connector/modules/chain_connector/src/chain/deal_changed.rs [hunk 28]
```diff
@@ -37,13 +38,27 @@ impl ChainData for DealChangedData {
 
     fn signature() -> Vec<EventField> {
         vec![
-            NotIndexed(ParamType::String), // appCID
+//            NotIndexed(ParamType::String), // appCID
+// app_cid
```
Comment (folex) [inline src/services/chain-connector/modules/chain_connector/src/chain/deal_changed.rs:42]: remove?

File: src/services/chain-connector/modules/chain_connector/src/chain/deal_changed.rs [hunk 29]
```diff
@@ -37,13 +38,27 @@ impl ChainData for DealChangedData {
...
+                // hash
+                ParamType::FixedBytes(32),
+            ]))
         ]
     }
 
     /// Parse data from chain. Accepts data with and without "0x" prefix.
     fn parse(data_tokens: &mut impl Iterator<Item = Token>) -> Result<Self, ChainDataError> {
-        let app_cid = next_opt(data_tokens, "app_cid", |t| t.into_string())?;
+//        let app_cid = next_opt(data_tokens, "app_cid", |t| t.into_string())?;
```
Comment (folex) [inline src/services/chain-connector/modules/chain_connector/src/chain/deal_changed.rs:54]: remove?

**Review by folex — APPROVED** [review]

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 25 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs]: well, I fixed CID encoding, but it still doesn't work with real chains. I created a task not to forget that it's broken https://linear.app/fluence/issue/NET-573/fix-deal-update-in-chain-connector

I kinda don't like that the tests on update are working and IRL it doesn't work

File: src/tests/decider-distro-tests-rs/tests/test.rs (hunk 26 above)
Comment (kmd-fl) [reply src/tests/decider-distro-tests-rs/tests/test.rs]: https://linear.app/fluence/issue/NET-574/make-a-range-of-blocks-for-polling-in-decider-configurable

**Comment by kmd-fl** [comment]
If you have unresolved questions, you are free to ask me directly.
